File: petshop_app/lib/models/message.dart
import 'user.dart';

class ChatMessage {
  final int id;
  final int senderId;
  final int receiverId;
  final String content;
  final String type;
  final bool isRead;
  final DateTime createdAt;
  final User? sender;

  ChatMessage({
    required this.id,
    required this.senderId,
    required this.receiverId,
    required this.content,
    required this.type,
    required this.isRead,
    required this.createdAt,
    this.sender,
  });

  factory ChatMessage.fromJson(Map<String, dynamic> json) {
    return ChatMessage(
      id: json['id'] ?? 0,
      senderId: json['sender_id'] ?? 0,
      receiverId: json['receiver_id'] ?? 0,
      content: json['content'] ?? '',
      type: json['type'] ?? 'text',
      isRead: json['is_read'] ?? false,
      createdAt: DateTime.parse(json['created_at'] ?? DateTime.now().toIso8601String()),
      sender: json['sender'] != null ? User.fromJson(json['sender']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'sender_id': senderId,
      'receiver_id': receiverId,
      'content': content,
      'type': type,
      'is_read': isRead,
      'created_at': createdAt.toIso8601String(),
      'sender': sender?.toJson(),
    };
  }
}

class Conversation {
  final int id;
  final int userId1;
  final int userId2;
  final ChatMessage? lastMessage;
  final int unreadCount;
  final DateTime updatedAt;
  final User? otherUser;

  Conversation({
    required this.id,
    required this.userId1,
    required this.userId2,
    this.lastMessage,
    required this.unreadCount,
    required this.updatedAt,
    this.otherUser,
  });

  factory Conversation.fromJson(Map<String, dynamic> json) {
    return Conversation(
      id: json['id'] ?? 0,
      userId1: json['user_id1'] ?? 0,
      userId2: json['user_id2'] ?? 0,
      lastMessage: json['last_message'] != null 
          ? ChatMessage.fromJson(json['last_message']) 
          : null,
      unreadCount: json['unread_count'] ?? 0,
      updatedAt: DateTime.parse(json['updated_at'] ?? DateTime.now().toIso8601String()),
      otherUser: json['other_user'] != null ? User.fromJson(json['other_user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id1': userId1,
      'user_id2': userId2,
      'last_message': lastMessage?.toJson(),
      'unread_count': unreadCount,
      'updated_at': updatedAt.toIso8601String(),
      'other_user': otherUser?.toJson(),
    };
  }
}

class SendMessageRequest {
  final int receiverId;
  final String content;
  final String type;

  SendMessageRequest({
    required this.receiverId,
    required this.content,
    this.type = 'text',
  });

  Map<String, dynamic> toJson() {
    return {
      'receiver_id': receiverId,
      'content': content,
      'type': type,
    };
  }
}

File: backend/app/models/message.py
from sqlalchemy import Column, Integer, String, Text, Boolean, DateTime, ForeignKey, Index
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func
from ..core.database import Base


class Conversation(Base):
    """对话表"""
    __tablename__ = "conversations"
    
    id = Column(Integer, primary_key=True, index=True)
    user1_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    user2_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    last_message_id = Column(Integer, ForeignKey("messages.id"), nullable=True)
    last_message_time = Column(DateTime, nullable=True)
    user1_unread_count = Column(Integer, default=0)
    user2_unread_count = Column(Integer, default=0)
    user1_deleted = Column(Boolean, default=False)
    user2_deleted = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    
    # 关系
    user1 = relationship("User", foreign_keys=[user1_id])
    user2 = relationship("User", foreign_keys=[user2_id])
    last_message = relationship("Message", foreign_keys=[last_message_id], post_update=True)
    messages = relationship("Message", back_populates="conversation", cascade="all, delete-orphan", 
                          foreign_keys="Message.conversation_id")
    
    # 索引
    __table_args__ = (
        Index('idx_conversation_users', 'user1_id', 'user2_id'),
        Index('idx_conversation_last_message_time', 'last_message_time'),
    )


class Message(Base):
    """消息表"""
    __tablename__ = "messages"
    
    id = Column(Integer, primary_key=True, index=True)
    conversation_id = Column(Integer, ForeignKey("conversations.id"), nullable=False)
    sender_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    receiver_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    message_type = Column(String(20), default="text")  # text, image, product, system
    content = Column(Text, nullable=False)
    related_id = Column(Integer, nullable=True)  # 关联的商品ID等
    is_read = Column(Boolean, default=False)
    is_deleted = Column(Boolean, default=False)
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())
    
    # 关系
    conversation = relationship("Conversation", back_populates="messages", foreign_keys=[conversation_id])
    sender = relationship("User", foreign_keys=[sender_id])
    receiver = relationship("User", foreign_keys=[receiver_id])
    
    # 索引
    __table_args__ = (
        Index('idx_message_conversation', 'conversation_id'),
        Index('idx_message_sender', 'sender_id'),
        Index('idx_message_receiver', 'receiver_id'),
        Index('idx_message_created_at', 'created_at'),
        Index('idx_message_unread', 'receiver_id', 'is_read'),
    )
